### PR TITLE
Optimize `isnan` (2-5x faster)

### PR DESCRIPTION
**Pull request by simonvandel**
## Which issue does this PR close?

Closes #.

## Rationale for this change

Apply same idea from https://github.com/apache/datafusion/pull/12881 to `isnan`.

## What changes are included in this PR?

- Add bench for is_nan
- Optimize implementation

## Are these changes tested?

Yes, with existing tests.

## Are there any user-facing changes?

Yes, faster `is_nan` function.

```
isnan f32 array: 1024   time:   [794.46 ns 795.95 ns 797.51 ns]
                        change: [-60.752% -60.453% -60.230%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

isnan f64 array: 1024   time:   [694.70 ns 698.25 ns 702.33 ns]
                        change: [-66.246% -66.029% -65.767%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) high mild
  4 (4.00%) high severe

isnan f32 array: 4096   time:   [2.5207 µs 2.5282 µs 2.5358 µs]
                        change: [-64.245% -64.126% -64.015%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

isnan f64 array: 4096   time:   [2.0779 µs 2.0812 µs 2.0847 µs]
                        change: [-74.140% -74.042% -73.950%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

isnan f32 array: 8192   time:   [4.9547 µs 4.9730 µs 4.9971 µs]
                        change: [-74.219% -74.059% -73.933%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

isnan f64 array: 8192   time:   [4.0478 µs 4.0562 µs 4.0652 µs]
                        change: [-82.993% -82.947% -82.900%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
```

### CODE REVIEW

File: datafusion/physical-expr/src/math_expressions.rs [hunk 1]
```diff
@@ -17,60 +17,27 @@
...
-                Some(a) => Some($FUNC(a)),
-                _ => None,
-            })
-            .collect::<$RETURN_TYPE>()
-    }};
-}
+use datafusion_common::Result;
 
 /// Isnan SQL function
 pub fn isnan(args: &[ArrayRef]) -> Result<ArrayRef> {
```
Comment (jonahgao) [inline datafusion/physical-expr/src/math_expressions.rs:31]: It looks like this function is unused, and the real one is in datafusion/functions/src/math/nans.rs.

Comment (simonvandel) [reply datafusion/physical-expr/src/math_expressions.rs]: You are absolutely right.
Changing the right function does indeed give a speed-up.

I have left out the cleanup of the `isnan` of math_expressions.rs out of this PR now.

**Review by jonahgao — APPROVED** [review]
LGTM, thanks @simonvandel

File: datafusion/physical-expr/src/math_expressions.rs (hunk 1 above)
Comment (jonahgao) [reply datafusion/physical-expr/src/math_expressions.rs]: PR #12917 to remove `math_expressions.rs`

**Comment by alamb** [comment]
Thanks @simonvandel and @jonahgao  -- I merged this PR up to resolve a conflict

**Pull request merged**